docs/CONTRIBUTING.md

```diff
@@ -1,11 +1,327 @@
+    - [OSTree tests](#ostree-tests)
+    - [Submitting a patch](#submitting-a-patch)
+    - [Commit message style](#commit-message-style)
+    - [Returning workflow](#returning-workflow)
+
+---
+
+## [Getting Started](#getting-started)
+
+Fork https://github.com/ostreedev/ostree, the run the following commands.
+
+```bash
+$ git clone https://github.com/<username>/ostree
+$ git remote add upstream https://github.com/ostreedev/ostree
+$ git checkout master
+$ git branch --set-upstream-to=upstream/master master
+```
+Make a branch from master for your patch.
+```bash
+$ git checkout -b <name-of-branch>
+$ git branch --set-upstream-to=upstream/master <name-of-branch>
+```
+
+## [Building OSTree](#building-ostree)
+
+### [Install Build Dependencies](#install-build-dependencies)
+
+```bash
+$ sudo dnf builddep ostree
+$ sudo dnf install @buildsys-build
+```
+
+### [OSTree Build Commands](#ostree-build-commands)
+
+```bash
+git submodule update --init
+./autogen.sh --prefix=/usr --libdir=/usr/lib64 --sysconfdir=/etc
+./configure --prefix=/usr
+make
+make install DESTDIR=/path/to/install/binary
+```
+
+`make install` will generate files for `/etc` and `/usr`
+
+For more information and issues with `./autogen.sh`, `./configure` or `make` see [Build Dependencies](/docs/misc/build-dependencies.md).
+
+## [Testing a Build](#testing-a-build)
+
+**TODO**: This sentense below needs more context and perhaps needs to be rephrased.
+
+Testing OSTree could interfere with the state of your machine. To avoid this, it is preferable to test and build OSTree inside a container or virtual machine.
+
+### [Testing in a Container](#testing-in-a-container)
+
+Docker is the preferred containerization tool. Follow this guide to [install Docker on your host machine](https://docs.docker.com/v17.09/engine/installation/linux/docker-ce/fedora/). You may need to follow this [post-installation guide for Docker](https://docs.docker.com/v17.09/engine/installation/linux/linux-postinstall/) if you would like to run Docker as a non-root user on a Linux based host machine.
+
+Save the contents of this **Dockerfile** somewhere on your machine:
+```bash
+# this pulls the latest fedora image
+FROM fedora:latest
+
+# install ostree dependencies
+RUN dnf update -y && \
+    dnf -y install dnf-plugins-core  && \
+    dnf -y builddep ostree  && \
+    dnf -y install make redhat-rpm-config && \
+    dnf clean all
+
+# clone your ostree repo
+RUN git clone https://github.com/<username>/ostree.git
+
+# pull patch-branch + set upstream
+RUN cd ostree && \
+    git pull origin <branch of your patch> && \
+    git checkout <branch of your patch> && \
+    git remote add upstream https://github.com/ostreedev/ostree.git
+
+WORKDIR /ostree
+```
+**Build the container**
+
+Run `docker build` in the same directory of the `Dockerfile` like so:
+
+```
+$ docker build -t ostree-fedora-test .
+```
+When this build is done, the `-t` option tags the image as `ostree-fedora-test`.  
```

Comment:
I'm OK with including this but I'd try really hard to avoid doing this myself as it's really, really slow for iteration to re-download the build dependencies each time.  This is why I like the "pet container" pattern more.

Reply:
I withdraw this comment; it's fine to demonstrate this, and it can make a lot of sense to have different build containers per project.